File: src/js/core/leetcode/addTwoNumbers.js
let ListNode = require("../support/custom_types/listNode").default

/**
 * Static Class methods for solving the Add Two Numbers Leetcode question
 */
class AddTwoNumbers{
    /**
     * Uses string manipulation to solve the algorithm
     * Transforms Node inputs to strings, to int, back to str
     * Iterates final string into a SingleLinkedNode as a return variable
     * @param l1 {SingleLinkedNode}
     * @param l2 {SingleLinkedNode}
     *
     * @see ListNode
     */
    static bruteForce(l1, l2){
        let [l1String, l2String] = ["", ""];
        /**
         * Creating strings out of the nodes
         */
        while (!(l1 == null && l2 == null)){
            if (!(l1 == null)){
                l1String += l1.val;
                l1 = l1.next
            }

            if (!(l2 == null)){
                l2String += l2.val;
                l2 = l2.next
            }
        }

        /**
         * Use to hold integers beyond Number.Max_Safe_Integer
         * @type {bigint}
         */
        const l1ReversedNum = BigInt(l1String.split("").reverse().join(""))
        const l2ReversedNum = BigInt(l2String.split("").reverse().join(""))
        /**
         * Parse the ListNode val as an array of integers
         * @type {string[]}
         */
        const retStringArray = String(l1ReversedNum + l2ReversedNum).split("").reverse()
        let previousNode = new ListNode(); // needed to keep track in memory as the retStringArray forEach is looping
        let ret = new ListNode();  // Initializing the ret ListNode

        /**
         * Use the Array for each prototype method
         */
        retStringArray.forEach((currentValue, index, arr)=>{
            if(index === 0 ){
                ret.val = Number(currentValue);
                previousNode = ret;
            } else {
                const currentNode = new ListNode(Number(currentValue));
                previousNode.next = currentNode;
                previousNode = currentNode;
            }

        })

        return ret
    }
    /**
     *
     * Uses a form of iteration to add a node to another node
     * @param l1 {SingleLinkedNode}
     * @param l2 {SingleLinkedNode}
     *
     * @see ListNode
     * @author Yaadata Abdalhalim
     */
    static optimized(l1, l2){
        let previousNode = null;
        let ret = new ListNode();
        let isHead = true;
        let carry = 0;
        /**
         * null, undefined and zero are all falsy values
         */
        while (l1 || l2 || carry){
            let currentValue = carry;

            if (l1){
                currentValue += l1.getVal;
                l1 = l1.getNext;
            }

            if (l2){
                currentValue += l2.getVal;
                l2 = l2.getNext;
            }

            if (currentValue >= 10){
                carry = 1;
                currentValue = currentValue - 10;
            } else {
                carry = 0;
            }

            if (isHead){
                ret.setVal = currentValue;
                previousNode = ret;
                isHead = false;
            }  else {
                const currentNode = new ListNode(currentValue)
                previousNode.setNext = currentNode;
                previousNode = currentNode
            }
        }

        return ret
    }
}

module.exports = {
    default: AddTwoNumbers
}

File: src/py/core/leetcode/add_two_numbers.py
from core.support.custom_types.list_nodes import SingleLinkedNode


class AddTwoNumbers:
    """
    Used to solved the Add Two Numbers leetcode medium question

    - see
        `test.test_leetcode.test_add_two_numbers.AddTwoNumbersTest`

    """
    @staticmethod
    def bruteforce(l1: SingleLinkedNode, l2: SingleLinkedNode) -> SingleLinkedNode:
        """
        Uses string manipulation to solve the algorithm
        Transforms Node inputs to strings, to int, back to str
        Iterates final string into a SingleLinkedNode as a return variable

        :param l1: SingleLinkedNode

        :param l2: SingleLinkedNode

        :return: SingleLinkedNode

        see
            `core.support.custom_types.list_nodes.SingleLinkedNode`
        """
        l1_str, l2_str = '', ''
        while l1 or l2:
            if l1:
                l1_str += str(l1.val)
                l1 = l1.next

            if l2:
                l2_str += str(l2.val)
                l2 = l2.next
        """
        1. Reverse each string
        2. Turn into integer
        3. Add the integers
        4. Create a string
        5. Reverse the string
        """
        ret_str: str = str(int(l1_str[::-1]) + int(l2_str[::-1]))[::-1]
        ret: SingleLinkedNode = SingleLinkedNode()
        previous_node = SingleLinkedNode()
        for char_index in range(0, len(ret_str)):
            if char_index == 0:
                ret.val = int(ret_str[char_index])
                previous_node = ret
            else:
                current_node = SingleLinkedNode(val=int(ret_str[char_index]))
                previous_node.next = current_node
                previous_node = current_node

        return ret

    @staticmethod
    def optimized(l1: SingleLinkedNode, l2: SingleLinkedNode) -> SingleLinkedNode:
        """
        Uses a form of iteration to add a node to another node

        :param l1: SingleLinkedNode

        :param l2: SingleLinkedNode

        :return SingleLinkedNode

        see `core.support.custom_types.list_nodes.SingleLinkedNode`
        """
        ret: SingleLinkedNode = SingleLinkedNode()  # return variable
        carry: int = 0
        previous_node: SingleLinkedNode = SingleLinkedNode()
        is_head: bool = True
        while l1 or l2 or carry:
            value: int = carry
            carry = 0
            """
            Add values only if l1 or l2 or carry exist
            """
            if l1:
                value += l1.val
                l1 = l1.next

            if l2:
                value += l2.val
                l2 = l2.next

            """
            Evaluate the remainder
            """
            quotient, remainder = divmod(value, 10)

            if quotient == 1:
                value = remainder
                carry += 1
            """
            Separate logic for head of list node and elements
            after the head
            """
            if is_head:
                is_head = False
                ret.val = value
                previous_node = ret
            else:
                current_node = SingleLinkedNode(val=value)
                previous_node.next = current_node
                previous_node = current_node

        return ret
